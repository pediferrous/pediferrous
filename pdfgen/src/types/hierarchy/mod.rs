//! General-purpose data structures that are built from the basic object types described in 7.3,
//! "Objects" and used throughout PDF technology. This subclause describes data structures for text
//! strings, dates, rectangles, name trees, and number trees.
//!
//! All of these data structures are meaningful only as part of the document hierarchy; they may
//! not appear within content streams. In particular, the special conventions for interpreting the
//! values of string objects apply only to strings outside content streams. An entirely different
//! convention is used within content streams for using strings to select sequences of glyphs to be
//! painted on the page.
//!
//! Reference: ISO 32000-2:2020 (PDF 2.0); page 114

pub mod catalog;
pub mod content;
pub mod cross_reference_table;
pub mod page;
pub mod page_tree;
pub mod primitives;
pub mod trailer;
