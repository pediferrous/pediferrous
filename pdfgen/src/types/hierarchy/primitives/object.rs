//! Definition of PDF object(trait).

use std::io::{self, Write};

use super::obj_id::ObjId;

/// The [`Object`] trait serves as a blueprint for all types that need to
/// provide a custom implementation for serializing or outputting their
/// structured data in a consistent manner.
pub trait Object {
    /// Writes the structured data of the object to the provided writer.
    fn write(&self, writer: &mut impl Write) -> Result<usize, io::Error>;

    /// Returns the [`ObjId`] associated with this object.
    fn obj_ref(&self) -> &ObjId;
}
