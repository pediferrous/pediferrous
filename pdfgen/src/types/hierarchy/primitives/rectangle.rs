use std::io::{Error, Write};

use super::unit::Unit;

/// Represents a point (pair of x and y coordinates) in default user space units.
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
pub struct Position {
    /// X coordinate in user space unit.
    x: Unit,

    /// y coordinate in user space unit.
    y: Unit,
}

impl Position {
    pub const fn new(x: Unit, y: Unit) -> Self {
        Self { x, y }
    }

    pub const fn from_mm(x: f32, y: f32) -> Self {
        Self {
            x: Unit::from_mm(x),
            y: Unit::from_mm(y),
        }
    }

    pub const fn from_units(x: f32, y: f32) -> Position {
        Self {
            x: Unit::from_unit(x),
            y: Unit::from_unit(y),
        }
    }
}

/// Rectangles are used to describe locations on a page and bounding boxes for a variety of
/// objects. A rectangle shall be written as an array of four numbers giving the coordinates of a
/// pair of diagonally opposite corners.
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
pub struct Rectangle {
    /// Lower left corner of the `Rectangle`.
    low_left: Position,

    /// Upper right corner of the `Rectangle`.
    top_right: Position,
}

macro_rules! gen_page_constants {
    ($($name:ident, $width:literal, $height:literal),* $(,)?) => {
        $(
        pub const $name: Self = Self::new(Position::from_mm(0.0, 0.0), Position::from_mm($width - 1.0, $height - 1.0));
        )*
    }
}

impl Rectangle {
    gen_page_constants! {
        A0_4, 1682.0, 2378.0,
        A0_2, 1189.0, 1682.0,
          A0, 841.0, 1189.0,
          A1, 594.0, 841.0,
          A2, 420.0, 594.0,
          A3, 297.0, 420.0,
          A4, 210.0, 297.0,
          A5, 148.0, 210.0,
          A6, 105.0, 148.0,
          A7, 74.0, 105.0,
          A8, 52.0, 74.0,
          A9, 37.0, 52.0,
         A10, 26.0, 37.0,
    }

    /// Create a new [`Rectangle`] with given [`Position`]s as its corners.
    pub const fn new(low_left: Position, top_right: Position) -> Self {
        Self {
            low_left,
            top_right,
        }
    }

    /// Encode and write this [`Rectangle`] into the provided implementor of [`Write`].
    pub fn write(&self, writer: &mut impl Write) -> Result<usize, Error> {
        let output = format!(
            "[{} {} {} {}]",
            self.low_left.x, self.low_left.y, self.top_right.x, self.top_right.y
        );
        let written = writer.write(output.as_bytes())?;
        Ok(written)
    }

    pub fn from_units(ll_x: f32, ll_y: f32, tr_x: f32, tr_y: f32) -> Self {
        Self {
            low_left: Position::from_units(ll_x, ll_y),
            top_right: Position::from_units(tr_x, tr_y),
        }
    }
}

impl From<(u32, u32, u32, u32)> for Rectangle {
    fn from((ll_x, ll_y, tr_x, tr_y): (u32, u32, u32, u32)) -> Self {
        Self::from((ll_x as f32, ll_y as f32, tr_x as f32, tr_y as f32))
    }
}

impl From<(f32, f32, f32, f32)> for Rectangle {
    fn from((ll_x, ll_y, tr_x, tr_y): (f32, f32, f32, f32)) -> Self {
        Self {
            low_left: Position::from_mm(ll_x, ll_y),
            top_right: Position::from_mm(tr_x, tr_y),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::types::hierarchy::primitives::unit::Unit;

    use super::Rectangle;

    #[test]
    fn new_rectangle() {
        let rect = Rectangle::from((24, 25, 42, 43));

        assert_eq!(rect.low_left.x, Unit::from_mm(24.0));
        assert_eq!(rect.low_left.y, Unit::from_mm(25.0));
        assert_eq!(rect.top_right.x, Unit::from_mm(42.0));
        assert_eq!(rect.top_right.y, Unit::from_mm(43.0));
    }

    #[test]
    fn output() {
        let rect = Rectangle::from_units(24.0, 25.0, 42.0, 43.0);

        let mut output = Vec::new();
        rect.write(&mut output).unwrap();
        let output = String::from_utf8(output).unwrap();

        insta::assert_snapshot!(output, @"[24 25 42 43]");
    }
}
