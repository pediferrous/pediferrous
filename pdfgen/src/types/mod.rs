pub mod hierarchy;
pub mod pdf_writer;
pub use hierarchy::page;

/// Common constants used when writing encoded PDF into a [`Write`] or [`PdfWriter`].
///
/// [`Write`]: std::io::Write
/// [`PdfWriter`]: super::pdf_writer::PdfWriter
pub mod constants {
    /// New line constant
    pub const NL_MARKER: &[u8] = b"\n";

    /// Single Space
    pub const SP: &[u8] = b" ";
}

/// Helper macro for counting the number of written bytes in multiple consecutive writes, where
/// each write returns a `Result<usize, std::io::Error>`
///
/// # Example
///
/// ```ignore
/// let mut writer = Vec::new();
/// let written = write_chain! {
///     writer.write(b"Hello"),
///     writer.write(b", World!"),
/// };
///
/// assert_eq!(written, 13);
/// ```
macro_rules! write_chain {
    ($($expression:expr),* $(,)?) => {{
        let mut written = 0;
        $(
            written += $expression?;
        )*
        written
    }}
}

pub(crate) use write_chain;

#[cfg(test)]
mod tests {
    #[test]
    fn write_chain() {
        use std::io::Write;

        let write_fn = || -> Result<usize, std::io::Error> {
            let mut writer = Vec::new();
            let written = write_chain! {
                writer.write(b"Hello"),
                writer.write(b", World!"),
            };

            Ok(written)
        };

        let written = write_fn().unwrap();
        assert_eq!(written, 13);
    }
}
