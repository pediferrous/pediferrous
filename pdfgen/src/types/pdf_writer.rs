//! Implementation of the [`PdfWriter`] wrapper.

use super::{
    constants,
    hierarchy::{
        cross_reference_table::CrossReferenceTable,
        primitives::{obj_id::ObjId, object::Object},
        trailer::WriteTrailer,
    },
};
use std::io::{self, Write};

/// A wrapper around any type that implements [`Write`], adding pdf specific functionality to keep a
/// clear and consistent CrossReferenceTable state
pub struct PdfWriter<W: Write> {
    /// Inner member, representing a type that implements [`Write`].
    inner: W,
    /// Current byte offset from the top of document, representing the current position of the `cursor`.
    current_offset: usize,
    /// CrossReferenceTable member, representing the current state of the cross_reference_table
    /// for the document
    cross_reference_table: CrossReferenceTable,
}

impl<W: Write> PdfWriter<W> {
    /// The PDF file begins with the 5 characters “%PDF–X.X” and byte offsets shall be calculated
    /// from the PERCENT SIGN.
    const PDF_HEADER: &[u8] = b"%PDF-2.0";
    /// The last line of the file shall contain only the end-of-file marker, %%EOF
    const EOF_MARKER: &[u8] = b"%%EOF";
    /// Marker indicating end of an object section
    const END_OBJ_MARKER: &[u8] = b"endobj";

    /// Creates a new [`PdfWriter`] instance.
    pub fn new(inner: W) -> Self {
        PdfWriter {
            inner,
            current_offset: 0,
            cross_reference_table: CrossReferenceTable::default(),
        }
    }

    /// Write the PDF documents header marker updating the `cursor`s byte offset with the number of
    /// bytes written.
    pub fn write_header(&mut self) -> Result<(), io::Error> {
        // Delegate the actual writing to the inner writer incrementing the current_offset to
        // reflect current `cursor` position.
        self.current_offset += self.inner.write(Self::PDF_HEADER)?;
        self.current_offset += self.inner.write(constants::NL_MARKER)?;

        Ok(())
    }

    /// Writes the object start marker(`X X obj`), following with the structured data of the object
    /// itself, finalizing with object end marker(`endobj`), ensuring correct CrossReferenceTable
    /// and cursor update.
    pub fn write_object(&mut self, obj: &dyn Object, obj_ref: &ObjId) -> Result<(), io::Error> {
        // Save the objects byte offset in the CrossReferenceTable.
        self.cross_reference_table.add_object(self.current_offset);

        // X Y obj\n
        self.current_offset += obj_ref.write_def(&mut self.inner)?;
        self.current_offset += self.inner.write(constants::NL_MARKER)?;

        // Delegate the actual writing to the inner writer.
        self.current_offset += obj.write(&mut self.inner)?;
        self.current_offset += self.inner.write(constants::NL_MARKER)?;

        // endobj\n
        self.current_offset += self.inner.write(Self::END_OBJ_MARKER)?;
        self.current_offset += self.inner.write(constants::NL_MARKER)?;

        Ok(())
    }

    /// Writes the cross reference table contents.
    pub fn write_crt(&mut self) -> Result<(), io::Error> {
        self.cross_reference_table.write(&mut self.inner)?;

        Ok(())
    }

    /// Writes the trailer for the PdfWriter's CRT.
    pub fn write_trailer(&mut self, root: ObjId) -> Result<(), io::Error> {
        self.cross_reference_table.write_trailer(
            &mut self.inner,
            self.current_offset,
            self.cross_reference_table.len(),
            root,
            self.cross_reference_table.offsets_hash()?,
        )?;

        Ok(())
    }

    /// Write the PDF documents EOF marker.
    pub fn write_eof(&mut self) -> Result<(), io::Error> {
        // Delegate the actual writing to the inner writer.
        self.inner.write_all(Self::EOF_MARKER)
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        types::hierarchy::primitives::obj_id::{IdManager, ObjId},
        PdfWriter,
    };

    use super::Object;

    struct Dummy(ObjId);
    impl Object for Dummy {
        fn write(&self, writer: &mut dyn std::io::Write) -> Result<usize, std::io::Error> {
            writer.write(b"FirstLine\nSecondLine")
        }

        fn obj_ref(&self) -> &crate::types::hierarchy::primitives::obj_id::ObjId {
            &self.0
        }
    }

    #[test]
    fn write_header() {
        let mut writer = Vec::new();
        let mut pdf_writer = PdfWriter::new(&mut writer);

        pdf_writer.write_header().unwrap();

        let output = String::from_utf8(writer).unwrap();

        insta::assert_snapshot!(
            output,
            @"%PDF-2.0"
        );
    }

    #[test]
    fn write_eof() {
        let mut writer = Vec::new();
        let mut pdf_writer = PdfWriter::new(&mut writer);

        pdf_writer.write_eof().unwrap();

        let output = String::from_utf8(writer).unwrap();

        insta::assert_snapshot!(
            output,
            @"%%EOF"
        );
    }

    #[test]
    fn write_object() {
        let mut writer = Vec::new();
        let mut pdf_writer = PdfWriter::new(&mut writer);
        let mut id_manager = IdManager::default();

        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();

        let output = String::from_utf8(writer).unwrap();

        insta::assert_snapshot!(
            output,
            @r"
        0 0 obj
        FirstLine
        SecondLine
        endobj
        "
        );
    }

    #[test]
    fn write_crt() {
        let mut writer = Vec::new();
        let mut pdf_writer = PdfWriter::new(&mut writer);
        let mut id_manager = IdManager::default();

        pdf_writer.write_header().unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer.write_crt().unwrap();
        pdf_writer.write_eof().unwrap();

        let output = String::from_utf8(writer).unwrap();

        insta::assert_snapshot!(
            output,
            @r"
        %PDF-2.0
        0 0 obj
        FirstLine
        SecondLine
        endobj
        1 0 obj
        FirstLine
        SecondLine
        endobj
        2 0 obj
        FirstLine
        SecondLine
        endobj
        3 0 obj
        FirstLine
        SecondLine
        endobj
        xref
        0 4
        0000000009 00000 n 
        0000000045 00000 n 
        0000000081 00000 n 
        0000000117 00000 n 
        %%EOF
        "
        );
    }

    #[test]
    fn write_trailer() {
        let mut writer = Vec::new();
        let mut pdf_writer = PdfWriter::new(&mut writer);
        let mut id_manager = IdManager::default();

        pdf_writer.write_header().unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer
            .write_object(&Dummy(id_manager.create_id()), &id_manager.create_id())
            .unwrap();
        pdf_writer.write_crt().unwrap();
        pdf_writer.write_trailer(id_manager.create_id()).unwrap();
        pdf_writer.write_eof().unwrap();

        let output = String::from_utf8(writer).unwrap();

        insta::assert_snapshot!(
            output,
            @r"
        %PDF-2.0
        0 0 obj
        FirstLine
        SecondLine
        endobj
        1 0 obj
        FirstLine
        SecondLine
        endobj
        2 0 obj
        FirstLine
        SecondLine
        endobj
        3 0 obj
        FirstLine
        SecondLine
        endobj
        xref
        0 4
        0000000009 00000 n 
        0000000045 00000 n 
        0000000081 00000 n 
        0000000117 00000 n 
        trailer
               << /Size 4
               /Root 4 0 R
               /ID [<ef11002f88c2f7ddd4db3d52963e3a91>
                  <ef11002f88c2f7ddd4db3d52963e3a91>
                  ]
               >>
        startxref
        153
        %%EOF
        "
        );
    }
}
