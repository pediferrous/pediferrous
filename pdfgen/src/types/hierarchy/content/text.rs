//! Implementation of PDF Text object.

use crate::types::{
    constants,
    hierarchy::primitives::{object::Object, rectangle::Position, string::PdfString},
};

/// Comment
#[derive(Debug)]
pub(crate) struct TextTransform {
    /// Comment
    position: Position,

    /// Comment
    size: u32,
}

/// Comment
#[derive(Debug)]
pub struct Text {
    /// Comment
    content: PdfString,

    /// Comment
    transform: TextTransform,

    /// Comment
    font_name: String,
}

impl Text {
    /// Comment
    pub const BT_MARKER: &[u8] = b"BT";
    /// Comment
    pub const ET_MARKER: &[u8] = b"ET";
    /// Comment
    pub const TF_OPERATOR: &[u8] = b"Tf";
    /// Comment
    pub const TD_OPERATOR: &[u8] = b"Td";
    /// Comment
    pub const TJ_OPERATOR: &[u8] = b"Tj";

    /// Comment
    fn builder() -> TextBuilder<false> {
        let txt = Self {
            content: PdfString::from(""),
            transform: TextTransform {
                position: Position::from_mm(0.0, 0.0),
                size: 12,
            },
            font_name: String::from_utf8_lossy(constants::DEFAULT_FONT).to_string(),
        };

        TextBuilder { inner: txt }
    }

    /// Comment
    fn expand(&mut self, content: impl Into<String>) {
        self.content.expand(content);
    }
}

impl Object for Text {
    /// Comment
    fn write_def(&self, writer: &mut dyn std::io::Write) -> Result<usize, std::io::Error> {
        Ok(pdfgen_macros::write_chain! {
            writer.write(Self::BT_MARKER),
            writer.write(constants::NL_MARKER),
        })
    }

    fn write_content(&self, writer: &mut dyn std::io::Write) -> Result<usize, std::io::Error> {
        Ok(pdfgen_macros::write_chain! {
            // /FName Size Tf
            writer.write(format!{"/{} {} ", self.font_name, self.transform.size}.as_bytes()),
            writer.write(Self::TF_OPERATOR),
            writer.write(constants::NL_MARKER),

            // posx posy Td
            writer.write(format!("{} {} ", self.transform.position.x, self.transform.position.y).as_bytes()),
            writer.write(Self::TD_OPERATOR),
            writer.write(constants::NL_MARKER),

            // (Text) Tj
            self.content.write_content(writer),
            writer.write(constants::SP),
            writer.write(Self::TJ_OPERATOR),
            writer.write(constants::NL_MARKER),
        })
    }

    /// Comment
    fn write_end(&self, writer: &mut dyn std::io::Write) -> Result<usize, std::io::Error> {
        Ok(pdfgen_macros::write_chain! {
            writer.write(Self::ET_MARKER),
            writer.write(constants::NL_MARKER),
        })
    }
}

/// Comment
pub struct TextBuilder<const IS_INIT: bool> {
    /// Comment
    inner: Text,
}

impl<const IS_INIT: bool> TextBuilder<IS_INIT> {
    /// Sets the position of the [`Text`] on a page.
    pub fn at(mut self, pos: Position) -> TextBuilder<true> {
        self.inner.transform.position = pos;
        TextBuilder { inner: self.inner }
    }

    /// Sets the content of the [`Text`].
    pub fn with_content(mut self, content: impl Into<String>) -> Self {
        self.inner.content = PdfString::from(content);
        self
    }

    /// Expands the content of the [`Text`] with the provided content.
    pub fn with_expanded_content(mut self, content: impl Into<String>) -> Self {
        self.inner.expand(content);
        self
    }

    /// Sets the size of the [`Text`].
    pub fn with_size(mut self, size: u32) -> Self {
        self.inner.transform.size = size;
        self
    }

    /// Sets the font of the [`Text`].
    pub fn with_font(mut self, font_name: impl Into<String>) -> Self {
        self.inner.font_name = font_name.into();
        self
    }
}

impl TextBuilder<true> {
    /// Comment
    pub fn build(self) -> Text {
        self.inner
    }
}

#[cfg(test)]
mod tests {
    use crate::types::hierarchy::primitives::object::Object;
    use crate::types::hierarchy::primitives::rectangle::Position;

    use super::Text;

    #[test]
    pub fn default_text() {
        let txt = Text::builder().at(Position::from_mm(0.0, 0.0)).build();

        let mut writer = Vec::default();
        let _ = txt.write_def(&mut writer);
        let _ = txt.write_content(&mut writer);
        let _ = txt.write_end(&mut writer);

        let output = String::from_utf8_lossy(&writer);
        insta::assert_snapshot!(output);
    }

    #[test]
    pub fn custom_text() {
        let txt = Text::builder()
            .with_content("This is")
            .with_expanded_content(" a custom text content.")
            .with_size(14)
            .with_font("CustomFnt")
            .at(Position::from_mm(0.0, 0.0))
            .build();

        let mut writer = Vec::default();
        let _ = txt.write_def(&mut writer);
        let _ = txt.write_content(&mut writer);
        let _ = txt.write_end(&mut writer);

        let output = String::from_utf8_lossy(&writer);
        insta::assert_snapshot!(output);
    }
}
