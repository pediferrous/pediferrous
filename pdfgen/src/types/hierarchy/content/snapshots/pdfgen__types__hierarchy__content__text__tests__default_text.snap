---
source: pdfgen/src/types/hierarchy/content/text.rs
expression: output
snapshot_kind: text
---
BT
/BiHDef 12 Tf
0 0 Td
() Tj
ET
